src/webgpu/util/texture/layout.ts

```diff
@@ -0,0 +1,194 @@
+    bytesPerRow * (size[1] / blockHeight - 1) + bytesPerBlock * (size[0] / blockWidth);
+  const byteLength = bytesPerSlice * (size[2] - 1) + sliceSize;
+
+  return {
+    bytesPerBlock,
+    byteLength: align(byteLength, kBufferCopyAlignment),
+    minBytesPerRow,
+    bytesPerRow,
+    rowsPerImage,
+  };
+}
+
+export function fillTextureDataWithTexelValue(
+  format: GPUTextureFormat,
+  outputBuffer: ArrayBuffer,
+  size: [number, number, number],
+  texelValue: ArrayBuffer,
+  bytesPerRow?: number,
+  rowsPerImage?: number
+): void {
+  const { blockWidth, blockHeight, bytesPerBlock } = kTextureFormatInfo[format];
+  assert(!!bytesPerBlock && !!blockWidth && !!blockHeight);
+  assert(bytesPerBlock === texelValue.byteLength);
+
+  const minBytesPerRow = (size[0] / blockWidth) * bytesPerBlock;
+  const alignedMinBytesPerRow = align(minBytesPerRow, kBytesPerRowAlignment);
+  if (bytesPerRow !== undefined) {
+    assert(bytesPerRow >= alignedMinBytesPerRow);
+    assert(isAligned(bytesPerRow, kBytesPerRowAlignment));
+  } else {
+    bytesPerRow = alignedMinBytesPerRow;
+  }
+
+  if (rowsPerImage !== undefined) {
+    assert(rowsPerImage >= size[1]);
+  } else {
+    rowsPerImage = size[1];
+  }
+
+  const bytesPerSlice = (bytesPerRow * rowsPerImage) / blockWidth;
+  const sliceSize =
+    bytesPerRow * (size[1] / blockHeight - 1) + bytesPerBlock * (size[0] / blockWidth);
+  const outputByteLength = bytesPerSlice * (size[2] - 1) + sliceSize;
+  assert(outputByteLength <= outputBuffer.byteLength);
+
+  const texelValueBytes = new Uint8Array(texelValue);
+  const outputTexelValueBytes = new Uint8Array(outputBuffer);
+  for (let slice = 0; slice < size[2]; ++slice) {
+    for (let row = 0; row < size[1]; row += blockHeight) {
+      for (let col = 0; col < size[0]; col += blockWidth) {
+        const byteOffset =
+          slice * rowsPerImage * bytesPerRow + row * bytesPerRow + col * texelValue.byteLength;
+        outputTexelValueBytes.set(texelValueBytes, byteOffset);
+      }
+    }
+  }
+}
+
+export function createTextureUploadBuffer(
+  device: GPUDevice,
+  format: GPUTextureFormat,
+  dimension: GPUTextureDimension,
+  size: [number, number, number],
+  texelValue: ArrayBuffer,
+  mipLevel: number = 0,
+  bytesPerRowIn?: number,
+  rowsPerImageIn?: number
+): {
+  buffer: GPUBuffer;
+  bytesPerRow: number;
+  rowsPerImage: number;
+} {
+  const { byteLength, bytesPerRow, rowsPerImage, bytesPerBlock } = getTextureCopyLayout(
+    format,
+    dimension,
+    size,
+    mipLevel,
+    bytesPerRowIn,
+    rowsPerImageIn
+  );
+
+  const [buffer, mapping] = device.createBufferMapped({
+    size: byteLength,
+    usage: C.BufferUsage.CopySrc,
+  });
+
+  assert(texelValue.byteLength === bytesPerBlock);
+  fillTextureDataWithTexelValue(format, mapping, size, texelValue, bytesPerRow, rowsPerImage);
+  buffer.unmap();
+
+  return {
+    buffer,
+    bytesPerRow,
+    rowsPerImage,
+  };
+}
+
+export function createTextureReadbackBuffer(
```

Comment:
what's the cpuData? Method name doesn't make it clear. Is it the expected data?

Reply:
I forget why I had this function, but I never use it, even in later patches, so it's removed